kube-core/src/labels.rs

```diff
@@ -0,0 +1,552 @@
+    pub fn extend(&mut self, exprs: impl IntoIterator<Item = Expression>) -> &Self {
+        self.0.extend(exprs);
+        self
+    }
+}
+
+impl Matcher for LabelSelector {
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool {
+        let selector: Selector = self.clone().into();
+        selector.matches(labels)
+    }
+}
+
+impl Matcher for Selector {
+    // Perform a match check on the resource labels
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool {
+        for expr in self.0.iter() {
+            if !expr.matches(labels) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl Matcher for Expression {
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool {
+        match self {
+            Expression::In(key, values) => match labels.get(key) {
+                Some(v) => values.contains(v),
+                None => false,
+            },
+            Expression::NotIn(key, values) => match labels.get(key) {
+                Some(v) => !values.contains(v),
+                None => true,
+            },
+            Expression::Exists(key) => labels.contains_key(key),
+            Expression::DoesNotExist(key) => !labels.contains_key(key),
+            Expression::Equal(key, value) => labels.get(key) == Some(value),
+            Expression::NotEqual(key, value) => labels.get(key) != Some(value),
+            Expression::Invalid => false,
+        }
+    }
+}
+
+impl Display for Expression {
+    /// Perform conversion to string
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Expression::In(key, values) => {
+                write!(
+                    f,
+                    "{key} in ({})",
+                    values.iter().cloned().collect::<Vec<_>>().join(",")
+                )
+            }
+            Expression::NotIn(key, values) => {
+                write!(
+                    f,
+                    "{key} notin ({})",
+                    values.iter().cloned().collect::<Vec<_>>().join(",")
+                )
+            }
+            Expression::Equal(key, value) => write!(f, "{key}={value}"),
+            Expression::NotEqual(key, value) => write!(f, "{key}!={value}"),
+            Expression::Exists(key) => write!(f, "{key}"),
+            Expression::DoesNotExist(key) => write!(f, "!{key}"),
+            Expression::Invalid => Ok(()),
+        }
+    }
+}
+
+impl Display for Selector {
+    /// Convert a selector to a string for the API
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let selectors: Vec<String> = self.0.iter().map(|e| e.to_string()).collect();
+        write!(f, "{}", selectors.join(","))
+    }
+}
+// convenience conversions for Selector and Expression
+
+impl IntoIterator for Expression {
+    type IntoIter = IntoIter<Self::Item>;
+    type Item = Self;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Some(self).into_iter()
```

Comment:
```suggestion
        std::iter::once(self)
```

Reply:
Unfortunately this is of type `iter::Once` which is not `Self::IntoIter`